p2p/ipld/plugin/nodes/nodes.go

```diff
@@ -0,0 +1,394 @@
+package nodes
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+
+	blocks "github.com/ipfs/go-block-format"
+	"github.com/ipfs/go-cid"
+	"github.com/ipfs/go-ipfs/core/coredag"
+	"github.com/ipfs/go-ipfs/plugin"
+	format "github.com/ipfs/go-ipld-format"
+	node "github.com/ipfs/go-ipld-format"
+	"github.com/lazyledger/nmt"
+	mh "github.com/multiformats/go-multihash"
+)
+
+const (
+	// Below used multiformats (one codec, one multihash) seem free:
+	// https://github.com/multiformats/multicodec/blob/master/table.csv
+
+	// Nmt is the codec used for leaf and inner nodes of an Namespaced Merkle Tree.
+	Nmt = 0x7700
+	// Sha256Namespace8Flagged is the multihash code used to hash blocks
+	// that contain an NMT node (inner and leaf nodes).
+	Sha256Namespace8Flagged = 0x7701
+
+	// DagParserFormatName can be used when putting into the IPLD Dag
+	DagParserFormatName = "extended-square-row-or-col"
+
+	// FIXME: These are the same as types.ShareSize and types.NamespaceSize.
+	// Repeated here to avoid a dependency to the wrapping repo as this makes
+	// it hard to compile and use the plugin against a local ipfs version.
+	// TODO: plugins have config options; make this configurable instead
+	namespaceSize = 8
+	shareSize     = 256
+)
+
+func init() {
+	registerNamespacedCodecOrPanic(
+		Sha256Namespace8Flagged,
+		"sha2-256-namespace8-flagged",
+		2*namespaceSize+sha256.Size,
+		sumSha256Namespace8Flagged,
+	)
+}
+
+func registerNamespacedCodecOrPanic(
+	codec uint64,
+	name string,
+	defaulLength int,
+	hashFunc mh.HashFunc,
+) {
+	if _, ok := mh.Codes[codec]; !ok {
+		// add to mh.Codes map first, otherwise mh.RegisterHashFunc would err:
+		mh.Codes[codec] = name
+		mh.Names[name] = codec
+		mh.DefaultLengths[codec] = defaulLength
+
+		if err := mh.RegisterHashFunc(codec, hashFunc); err != nil {
+			panic(fmt.Sprintf("could not register hash function: %v", mh.Codes[codec]))
+		}
+	}
+}
+
+// sumSha256Namespace8Flagged is the mh.HashFunc used to hash leaf and inner nodes.
+// It is registered as a mh.HashFunc in the go-multihash module.
+func sumSha256Namespace8Flagged(data []byte, _length int) ([]byte, error) {
+	isLeafData := data[0] == nmt.LeafPrefix
+	if isLeafData {
+		return nmt.Sha256Namespace8FlaggedLeaf(data[1:]), nil
+	} else {
+		return nmt.Sha256Namespace8FlaggedInner(data[1:]), nil
+	}
```

Comment:
This replaces the changes in go-multihash (here: https://github.com/lazyledger/go-multihash/pull/1/files#diff-433d03ee232cd8a6349216dc514204d0c1313aabad7445491142de300443ec90)